### ISSUE TITLE
ENH: stats: univariate distribution meta-issue

### ISSUE THREAD

**Issue by mdhaber** [scipy/scipy#15928]
The existing univariate distribution infrastructure is an extremely valuable feature of SciPy, and it offers two main ways of interacting with distributions.

1. `stats.norm.pdf(2, loc=0, scale=1)` calls the `pdf` method of the `norm` object that is instantiated at import time
2. `stats.norm(loc=0, scale=1).pdf(2)` uses the `__call__` method of the `norm` object to create a "frozen" instance of the distribution, then calls that object's `pdf` method.

There are two major advantages of approach 1 (and, to be fair, I'll be happy to add others as they come up in the comments below).

- It is a bit more natural than approach 2 for very simple invocations. For example, it may be more natural to think of the standard normal PDF a function and call `stats.norm.pdf(2)` rather than thinking of (and instantiating) a standard normal object and invoking its `pdf` method (`stats.norm().pdf(2)`).
- It is the more common approach in existing code.

However, there are several important disadvantages.

- All ~125 distributions are instantiated upon import, increasing import time.
- Generating complete, distribution-specific documentation is impractical (gh-10328)
- Parameters are processed every time a method is called, increasing overhead
- gh-13986

These could easily be avoided if `norm` were simply a _class_ (rather than an instance of the `norm_gen` class) and the user were required to instantiate that class (e.g. `norm()`) before invoking methods. This has been noted before; for example, see gh-12133 for an earlier discussion of the history and pros/cons of the two approaches.

There are also several issues that have been reported over the years that may not indicate _inherent_ disadvantages of the existing infrastructure, but do suggest that some major upgrades are needed. There are several features and limitations of the existing infrasture (inflexibility w.r.t. adding/removing method arguments) that suggest a need to start fresh, reimagining the infrastructure with these enhancements in mind rather than adding patches to the existing patchwork. 

* gh-2509
* gh-4538
* gh-4598 
* gh-6038
* gh-7311
* gh-9439
* gh-11465

This enhancement proposal is for a new univariate distribution infrastructure. Of course, we'd want to discuss and study other possibilities to make sure we get things right, but it might look something like this.

* Distribution families are represented by classes. The names are typically the same as existing distribution objects except in CamelCase (`norm` -> `Norm` or `Normal`, `loguniform` -> `LogUniform`).
* Distributions are instantiated before using methods. The distribution class documentation describes distribution shape parameters, which are passed as keyword-only arguments to the class `__init__` method (only). Naturally, this makes alternative parameterizations easy to support. Default support for `loc` and `scale` is provided, but distributions can disable them as needed. (Contrast against gh-15889.)
* For the most part, public methods include those that are currently available with minor changes (e.g. the name of `nnlf` is corrected to `nllf`). 
* The shape parameters used to initialize the distribution can be used as guesses to the `fit` method, which modifies the shape parameters of the distribution. Methods are available to extract the shape parameters.
* There are new methods for calculating the `mode` and partial derivatives of the PDF (useful for random variate generation and fitting parameters to data).
* As in the existing infrastructure, public methods are rarely overridden by subclasses; instead, subclasses almost exclusively override private methods. However, the new infrastructure is more strict about this distinction (e.g. there is a new `_fit` method so that fit overrides do not have to repeat boilerplate code) and there are different versions of private methods that can be overridden depending on (for example) whether the developer is able to vectorize with respect to shape parameters or would prefer to rely on a default loop.
* Methods have parameters that enable the user to control tradeoffs between precision/reliability and speed. For example, the user can opt out of expensive data validation checks and opt into special accuracy improvements (e.g. gh-2509).
* The infrastructure provides a standard mechanism for switching to special case implementations (e.g. t distribution to normal distribution for df=inf) rather than requiring developers to use `np.where`/`np.lazywhere`.
* Distributions "know" more about their parameters (e.g. like `_shape_info`, which was recently added to the existing infrastructure). Based on this information, parameter documentation and input validation are generated automatically.
* The infrastructure natively supports truncated support, wrapped domains, and complex inputs.

The new infrastructure and distributions would be widely advertised, and documentation and existing code would transition to the new infrastructure. Existing infrastructure/distributions would be supported without deprecation warnings until two releases before SciPy 2.0, at which point they would be moved to a new package that is released once (without planned maintenance).

@tupui @ev-br @tirthasheshpatel @chrisb83

**Comment by tupui** [comment]
I am +100 on this proposal. There are too many issues with the current infrastructure and it's very hard to maintain. Starting of with a clean slate will definitely be a good thing.

**Comment by chrisb83** [comment]
Thanks for writing this clear proposal. I will take a closer look at the issue 12133 to learn more about the background of this topic. Overall it makes a lot of sense to me.

At this point, I have just one comment:

> There are new methods for calculating the mode and partial derivatives of the PDF (useful for random variate generation and fitting parameters to data).

Maybe `dlogpdf` (derivative of logpdf) could be considered, too: potentially it is useful to express `nllf` and I think some of UNU.RAN methods use it (I need to check in more detail). So not very important at this stage, but let's keep it in mind.

**Comment by josef-pkt** [comment]
I jsut saw this

This will be difficult.
Many of the problems are orthogonal to class versus instance, e.g. vectorizing when support depends on parameters is never trivial.

Main design issue: stateless classes and instances versus (some) fixed parameters (fixed attributes of a class defined in `__init__`.
Both are convenient and very useful for different cases and distributions. 
Allowing both in a generic way looks to me very difficult.

**Comment by tupui** [comment]
@josef-pkt to

> Allowing both in a generic way looks to me very difficult.

One of the main point of this proposal is to remove the concept of frozen distributions. There would be only one way to play with distributions: which is using a class instance. (Or a factory, something else. We need to take some time to play with different APIs' architecture and investigate that with various distributions.) The current duality is too problematic/buggy and complicate the user API (and maintenance). We want to have a single point of entry for shape parameters, the init of the class and then you just do what you want with the distribution.

The point is that we only provide a single Pythonic and clean way to do things user side. For users using the class-based approach, it should not deviate much more. But in the background things would be extremely different with a more manageable infrastructure which can be evolved (e.g. right now adding a QMC sampler would be a pain), documented properly, etc.

**Comment by tirthasheshpatel** [comment]
I too am +1 for overhauling the frozen distribution infrastructure. To minimize breaking code, it would be good to keep all the distributions in the `stats` module as-is and declare them legacy.

> Main design issue: stateless classes and instances versus (some) fixed parameters (fixed attributes of a class defined in `__init__`.

@josef-pkt I think simply initializing a new class for different parameters to achieve the stateless class behavior isn't too bad. We can also add static methods in the new distribution classes that offer the stateless classes syntax. But not having it at all also doesn't seem like an unreasonable trade-off.

**Comment by tupui** [comment]
> @josef-pkt I think simply initializing a new class for different parameters to achieve the stateless class behavior isn't too bad. We can also add static methods in the new distribution classes that offer the stateless classes syntax. But not having it at all also doesn't seem like an unreasonable trade-off.

Just to be clear. The goal is to start with a clean slate, meaning the API can break things. We should only consider adding things if they make sense in terms of architecture. Not to comply with old behaviours. I would not add any mechanism to have a stateless distribution.

**Comment by josef-pkt** [comment]
Instantiating a new class for different parameters sounds like a large overhead.
What's the overhead of creating  a new instance each time when I simply want to call norm.pdf(loc, scale), or calling logpdf inside fit?

On the other hand, there are some distributions that would benefit from precomputing some constants for method calls with the same distribution parameters.

(In statsmodels we use the equivalent of stateless, semi-frozen (some parameters fixed in `__init__`) and frozen distributions.)

**Comment by josef-pkt** [comment]
I will see it when you have a prototype

tupui didn't like it when I moved most parameters to the methods in copulas so we can use it as a semi-frozen or unfrozen distribution. But that looks the better design to me for building models on top of the distributions.

In statsmodels, all (almost all) parameters that are estimated in models are arguments in model methods. In some cases it gets messy when we need to change additional model attributes that are not method arguments.
(But our models are a lot more complicated than a scipy distribution.)

One pattern that I use is to allow both, set some parameters in `__init__` as attributes, and allow method arguments to override them. However, this is fragile if there are extra model attributes that are computed from the changed attributes.

**Comment by josef-pkt** [comment]
just one more usecase that I remembered

finite difference derivative approximations need to compute the logpdf at different distribution parameters in one function.
extreme case:
A model in statsmodels that uses directly the scipy distribution for logpdf (in most cases we have our own functions for that, but not for all)
Uing a newton optimizer, requires logpdf and first and second derivatives (score and hessian). If they are not available in explicit form, then score and hessian are computed using numerical derivatives (finite difference or complex step).
So, each iteration of the optimizer will call the logpdf/loglike many times with different parameter values.

On the other hand, when we have estimated the model, then we have a `get_distribution` method that transforms the estimated parameters into the parameterization of the scipy distribution and returns a frozen distribution to the user, which then can be used to compute pdf, cdf, rvs based on the estimated model and parameters.

**Comment by tupui** [comment]
Following your example. You could write a small wrapper that would re-use a given class instance and just vary attributes (shape parameters). It might even be faster than instantiating a frozen distribution every time you want to update some parameters.

I don't see any issues there.

> tupui didn't like it when I moved most parameters to the methods in copulas

What I did not like at the time was the use of `*args, **kwargs` which IMO is not a good design as it prevents easy inspection of method options/params, documentation etc. Seems like the situation improved a bit-from my perspective-with more descriptive args.

**Comment by josef-pkt** [comment]
> You could write a small wrapper that would re-use a given class instance and just vary attributes (shape parameters). It might even be faster than instantiating a frozen distribution every time you want to update some parameters.

But that's not needed right now because the scipy.stats distributions have stateless methods, i.e. we don't need to create a new instance when changing parameters, we always use the same instance.

If we loose that, then we cannot use any scipy distribution without fully wrapping it for the estimation models, and carry around our own case specific instances.
And users that want to use our GenericLikelihoodModel will also need to have a wrapper for the scipy distributions.

**Comment by tupui** [comment]
I don't believe that having to modify the shape parameters of a distribution is the common use of distributions. I am not saying this is not useful, but that it does not represent the majority of usages. I do not think we should maintain such mechanism as it's quite straightforward to do on user/downstream side.

```python
dist = toto(a, b)

for ai, bi in ... :
    dist.a, dist.b = ai, bi
    dist.pdf(x)

    # instead of
    toto.pdf(x, a, b)
```

**Comment by josef-pkt** [comment]
> dist.a, dist.b = ai, bi
    dist.pdf(x)

If you allow that, then you don't gain anything at all in the current implementation.
If it is allowed, then the distributions still need to be stateless except for the distribution args, i.e. no precomputed attribures, or you have to wip and update them each time.

For users, this is pain without gain.

> I don't believe that having to modify the shape parameters of a distribution is the common use of distributions.

Shape parameters, but also loc and scale.
It's one of the main usecases in statsmodels.

other examples
confidence intervals for parameters that sometimes need rootfinding, similarly profile likelihood (confint)
power and sample size computation where scale or shape parameters change. (generic sample size computation again requires rootfinding for different parameters. In statsmodels those functions directly use scipy distributions, mainly norm, t, f, chi2 and their non-central versions, but also beta, binomial, poisson, gamma, ... for countmodels)

statsmodels uses frozen distributions only as user convenience. Users don't have to carry around the distribution args when providing or receiving a distribution instance.

Bayesian or Monte Carlo applications require mixture distributions where parameters change. Very often those cannot be fully vectorized and we cannot avoid many calls to distribution methods with different parameters (shape, loc, scale).

**Comment by NeilGirdhar** [comment]
This is an excellent proposal.  Some thoughts:

If you're redesigning anyway, you may want to take a look at my [efax](https://github.com/NeilGirdhar/efax) library, which was designed with machine learning in mind.  In particular,
- It stores vectorized parameters and all operations are vectorized.  This is much more efficient when working with multiple parameter values.
- It has a pair of parametrizations for all exponential families: the natural and expectation parametrization.  These are useful in machine learning.
- It can be queried about the natural of the parameters: their support, their shape (whether they are scalar, vector, or matrix parameters), and other invariants (are they symmetric or Hermitian matrices), etc.

There needs to be support for integer, real, and complex parameter types.  I discuss many practical redesign ideas in #7311.

The design of distributions with bounded support is suboptimal due to making all distributions in scipy location-scale families.  They should not be location scale families.  I discuss this in #6038.

**Comment by tupui** [comment]
Thank you @NeilGirdhar for the information. We will have a look. Note that we would not be able to use any Jax code as we don't want to add another dependency.

**Comment by NeilGirdhar** [comment]
> we would not be able to use any Jax code as we don't want to add another dependency.

Of course!  I only show it as illustration of an alternative design from which you may be able to take some ideas (parameter inspection, vectorization, multiple parametrizations, etc.)

**Comment by NeilGirdhar** [comment]
> I don't believe that having to modify the shape parameters of a distribution is the common use of distributions. I am not saying this is not useful, but that it does not represent the majority of usages. I do not think we should maintain such mechanism as it's quite straightforward to do on user/downstream side.

I agree with you @tupui.  At least with machine learning, the number of instantiations is very small provided that you support vectorization in all distributions (which you don't yet).

I agree that this form `toto.pdf(x, ai, bi)` is overcomplicated to my eyes.

**Comment by scottshambaugh** [comment]
I expect this overhaul would fix thing like [this](https://stackoverflow.com/questions/60981879/creating-scipy-stats-random-variable-subclass-does-not-result-in-expected-object) which I am still working around, so a big +1 in general.

**Comment by NeilGirdhar** [comment]
What is the process for this redesign?  Is there somewhere the interface is being discussed?  I'd like to follow along the discussion if possible.

I'm not sure what the usual process is, but it might be illustrative to design the interface for a single distribution (say `MultivariateNormal`) as an archetype.  That way the many people watching this issue can ensure that all of the linked problems have been eliminated.

**Comment by tupui** [comment]
Thank you for asking @NeilGirdhar. We did not formally start yet as this would be part of a grant project. We will let everyone know when/if we have the funding and how we can proceed. In any case, everything will be public and we will seek everyones' input 😃

**Comment by mdhaber** [comment]
To clarify, this issue is about univariate distributions. Multivariate normal is a multivariate distribution. Multivariate distributions could use more of an infrastructure than they have make them more consistent, but I think it would be more challenging than it's worth to share innards with the univariate distributions. For those, please continue the discussion in gh-16278.

**Comment by NeilGirdhar** [comment]
>  I think it would be more challenging than it's worth to share innards with the univariate distributions. 

I agree with you about the innards.

Still, if you're going to clean up the interface, it might be nice from a user point of view to provide the same pretty interface for multivariate distributions--even if the innards are different.

**Comment by fancidev** [comment]
Is there a place to see the current progress of this branch of development?

**Comment by NeilGirdhar** [comment]
I was just reminded of this issue when I saw the recent release of scipy that now supports the Array API.  As soon as Jax supports the Array API, I will port my distribution library [efax](https://github.com/NeilGirdhar/efax) to the Array API.  Efax was designed with many of the above shortcomings in mind.  It supports:
* automatic broadcasting on *all* parameters,
* multiple parametrizations for each distribution (in particular, the natural and expectation parametrizations of exponential families),
* introspection for all parameters, and
* complex parameter support.

It's also significantly simpler to add distributions using efax's framework that scipy's framework.

I realize it would be a lot of work, but what are people's thoughts on merging all of the functionality from scipy's distribution functions with the efax's framework?

**Comment by tupui** [comment]
Hi @NeilGirdhar @fancidev 👋 

We are making good progress. The active stats maintainers are involved.

We have a functional draft which is almost feature complete and addresses all issues we wanted to cover. Including Array API support 😃 (I worked on the I infrastructure to allow that to happen on SciPy in the first place). 

At that time though we are not ready for receiving external comments nor review of our design and implementation.

Jax is not considered at the moment. Jax itself is not Array API compatible and my understanding is that the philosophy from their team is different. I can have look at your package to see if there is something we are missing.

**Comment by NeilGirdhar** [comment]
Awesome, looking forward to seeing the new design!

> I can have look at your package to see if there is something we are missing.

Cool, sounds good!

(I only mentioned Jax to say that my timeline will follow theirs.)

**Comment by fancidev** [comment]
Thanks for the update @tupui.

A small suggestion for the `rv_continuous.fit` method for your consideration in the new framework: the `method` parameter should have default value `None`,  with the below semantics: If `method` is `None`, a suitable consistent estimator is used, with MLE preferred.

This (breaking) change allows alternative methods to be used when MLE is known to be inconsistent, such as when estimating the location of a gamma distribution with shape less than one. It also allows for example an unbiased estimator of the normal variance, which is probably more commonly used than the biased MLE.

**Comment by fancidev** [comment]
Another really small (but breaking) suggestion: the first parameter of `ppf` and `isf` of `rv_continuous` should read `p` not `q`, because it’s a probability rather than a quantile. (This change is breaking unless the parameter is made position-only, which unfortunately is also breaking.)

**Comment by tupui** [comment]
Thank you for the suggestion @fancidev.

To be clear for everyone following this work: first of all, thank you 😃 we really appreciate that, do read and take into consideration what users tell us. Our ultimate goal is to provide something useful for the community.

Now, the outcome of this work has never been to upgrade the current infrastructure but instead to start from a clean slate. Meaning that we are rewriting a whole new infra from the ground-up, learning from all past pain points and addressing our wish lists.

I cannot share more than that for now as we are still evaluating all options (including eventual upgrade or compatibility options). I hope everyone sees this as a good thing in the sense that we are doing our best to give all ideas their best shot and don't constrain ourselves to pre-existing practices. One of the main goal of this project was to challenge how we go about univariate distributions and this is precisely what we are doing.

What I can share though is that we will very open up a formal RFC (request for comments) period in the near future.

**Comment by NeilGirdhar** [comment]
> What I can share though is that we will open a formal RFC (request for comments) period in the near future.

That's really great to hear and very considerate to users.  Thank you!

(I think some people's comments were motivated by a fear that there wouldn't be such a period, which might make any oversights difficult to change.)

We are all eagerly looking forward to playing with the new design.

**Comment by fancidev** [comment]
Thanks for sharing! Looking forward to the RFC!

**Comment by NeilGirdhar** [comment]
@tupui @mdhaber I just want to say congratulations on an absolutely beautiful design!  This write-up with the links to the relevant is really great.

As I said when we spoke on Monday, there's a lot of things I love about this design and I wanted to voice my support for your design choices.  In particular:
* Absolutely love the "transformers" (wrapping, truncation, mixtures, doubling, etc.)   Beautiful!
* Love how there's exactly one way to do things (e.g., evaluating a pdf is done by constructing the object, then calling a method; there is no second pathway).
* Having worked on distribution changes in the past, I love the simplicity with which a new distribution can be updated.

I have a few questions and comments (and I may add more here as I think of them so as not to flood the thread).

---

Have you considered making the distribution naming a little bit more consistent with the Array API library and in particular [your other proposed extension](https://github.com/data-apis/array-api/issues/725).  In that proposal, I really liked the use of underscores that make code easier to read.  After all, code is read many more times than it's written.

For this proposal, it might make sense to use some slightly longer names: `log_pdf`, `log_entropy`, `inverse_cdf`, `complementary_cdf`.  These are longer, but they can save on explanatory comments.

---

Which methods use the RNG?  Right now, there's only `sample`?  Would it make sense to just pass in the RNG into those methods?  Storing RNGs may end up being very weird in Jax.

---

I love the inheritance hierarchy.  Have you considered adding a base `Distribution` class above `ContinuousDistribution` and `DiscreteDistribution`?  This could just have `support`.

---

I think we talked a bit about parameter introspection, and you said it was more of a future plan, which makes sense.

---

Anyway, I'm really excited about the new design!

**Comment by fancidev** [comment]
The proposed refactoring looks really cool!

Skimming through the notebook and documentation, I only have some “polishing” type of comments which is not important for initial release. Good stuff!

**Comment by fancidev** [comment]
A few comments about the design:

- Would it make sense to make a distribution object _immutable_? This should simplify reasoning. For example, that `fit` mutates the object feels weird to me.
- Would it be appropriate to detach `tol`, `cache`, and `rng` from the object? It helps to make the object lean. `tol` and `rng` are probably better specified at the call site. `cache` could be implemented by a `CachedDistribution` wrapper (which I don’t expect many to use?)
- Does the infrastructure allow the potential to support mixed continuous and discrete distribution?

**Comment by fancidev** [comment]
> For this proposal, it might make sense to use some slightly longer names: `log_pdf`, `log_entropy`, `inverse_cdf`, `complementary_cdf`. These are longer, but they can save on explanatory comments.

+1 on this one. I'd personally prefer `inv_` to `inverse_` and `sf` to `ccdf`, but the more verbose versions are definitely easier to draw attention than the subtle `iccdf` :)

> Also, I suggest moving the `fit` function out of the distribution and making it a free function. After all, it doesn't seem to rely on private members, so by OO principles, it should be a free function. You can provide a more efficient fit mechanism for exponential families.

+1 on having the free `fit` function. `scipy.stats` already has one (that fits to data), so just need to find a way to extend its functionality. I still find it useful to allow individual distributions to provide specialised fitting functions, e.g. via something like `_fit_formula`. The benefit is efficiency, accuracy, robustness, etc.

> I love the inheritance hierarchy. Have you considered adding a base `Distribution` class above `ContinuousDistribution` and `DiscreteDistribution`?
+1, though it can always be added later without affecting the existing interface. As a proof of concept, I'd also like to insert `InteriorContinuousDistribution` class (gh-19946) in between `Distribution` and `ContinuousDistribution` to support continuous distributions with a probability mass at boundary, such as compound Poisson Gamma (Tweedie).

> Anyway, I'm really excited about the new design!
Me too!

**Comment by tupui** [comment]
Thanks all, I am adding a quick note here just to say that we are reading. We just want to let people have a look first before answering to hopefully not bias too much the conversations 😃

After a week or so, I will sum up the discussions and make a reply to address the suggestions.

**Comment by nimish** [comment]
I'd add:

* Add cumulant support, since going between them and moments is sometimes useful generally and is essentially required to compute moments in closed form for sum distributions
* Mixed continuous and discrete distributions are rare but do exist https://en.wikipedia.org/wiki/Tweedie_distribution . It'd make more sense to me for the distribution infrastructure to be as support-independent as possible, even down to being generic over it.
* Type annotations, it's confusing to tell what function params actually mean when implementing them

**Comment by rlucas7** [comment]
Thanks everyone who worked on this behind the scenes and to those who've provided  comments. My 2 cents

 On moments: 
> * Add cumulant support, since going between them and moments is sometimes useful generally and is essentially required to compute moments in closed form for sum distributions

Cumulants and moments (eg mapping between the two) is useful for both continuous and discrete distributions in applications. [Cornish Fisher](https://en.wikipedia.org/wiki/Cornish%E2%80%93Fisher_expansion) comes to mind, as does [Saddlepoint expansions](https://en.wikipedia.org/wiki/Saddlepoint_approximation_method) which are sometimes formulated this way. 

For discrete distributions (and maybe sometimes continuous) it is often easiest to write a distributions moments by working in the factorial basis and converting to raw 
moments.
 Two methods which could do this would use the Stirling numbers of first and second kinds: 


$$ x^n = \sum_{k=1}^{n}{n \brace k} (x)_k  $$

and 

$$(x)_{n}=\sum_1^n {n \brack k} x^{k}$$.

(for whatever reason the github latex isn't rendering the latter change of basis correctly and the way I was able to get it to render was by removing the `k=` in the `\sum`) 

so converting back and forth is a change of basis (but can make computations easier). 
If we can leverage these change of basis it might be a nice use case. 

I have WIP CR for First kind Stirling and Second kind Stirling was merged a release or so ago. These codes are in scipy.special and support broadcasting. 

Example common distributions where this is handy (Poisson, Binomial, Negative Binomial), all three have simpler arbitrary positive integer moment representations that are much cleaner than the raw moments. 

> * Mixed continuous and discrete distributions are rare but do exist https://en.wikipedia.org/wiki/Tweedie_distribution . It'd make more sense to me for the distribution infrastructure to be as support-independent as possible, even down to being generic over it.

IIUC @mdhaber has noted this is planned, at least I'm assuming that's what is meant by mixture distributions (and not variance scale-mixture families which is a wholely different class of distributions). 

> * Type annotations, it's confusing to tell what function params actually mean when implementing them

+1 these would be nice to have. 

One thing I haven't noticed mentioned (possible I missed it though) is which methods are required and which aren't? (when implemented a new distribution) 

There are some distributions which have inverse CDFs easily represented but not CDFs and some which have nice algorithms for Sampling but not other things like easy CDFs (example is the Dickman or Goncharov distribution-this one has some nice CFTP sampling approaches).  Tweedie is maybe another one in this domain for certain values of `p` parameter and Polya-Gamma is another. 

These are maybe less important to the generalist scipy user but for infrastructure we might consider them, if nothing else at least to determine and document whether they or distributions like them would or would not be implementable using planned infrastructure.  

(apologies in advance if the 'public' interface aspect is noted somewhere in thread above and I missed)

**Comment by tupui** [comment]
Thanks everyone for your comments and support, we really appreciate you taking the time.

I am trying in the following to address the comments. Feel free to add more comments if I forgot anything.

Based on the overall positive response, thank you again, @mdhaber is going to open a PR. I will post another message on the forum for it too.

---

> Have you considered making the distribution naming a little bit more consistent

We did consider alternative naming schemes (we did a poll between us too) and the API we have now is the solution we came up with. Being slightly more descriptive with `complementary_cdf` would be nice, but then for consistency we would need `inverse_log_complementary_cdf` which seems too long for us. We also try to avoid adding small things like `inv` or `comp` as we had issues in the past with acronyms becoming slurs. Just a precaution here.

> Which methods use the RNG? Right now, there's only sample? Would it make sense to just pass in the RNG into those methods? 

We had quite some back and forth there too and are still discussing this point. Some of us thing we should not allow passing a RNG in methods and just allow having it at the init. Right now it does both. This is still a point open for discussions, we will have a second look.

> Also, I suggest moving the fit function out of the distribution and making it a free function. After all, it doesn't seem to rely on private members, so by OO principles, it should be a free function.

Yes and no. Things like staticmethod for instance are here to group methods under a given "name space" to have a logical grouping. We will think this once more.

But yes, after discussing this. We are going to remove the fit method from the initial API. And we will think about how to follow up.

>  Have you considered adding a base Distribution class ... It would serve as a type annotation for generic functions that accept distributions

Typing is not going to be part of this first iteration. But it's definitely something that we are going to look into in follow ups. (I am myself very much interested into typing. There are just a lot of initiatives right now happening on that front and waiting a little bit more might help us get there.)

> distribution object immutable

In the previous design, there was 2 ways to create a distribution: a classical class instance, and a frozen one. This design lead to a lot of confusion in the community and this is precisely one of the pain point this new design addresses.

We want to allow class mutability to be able to change parameters (in a controlled manner with proper checks). This will be important for some downstream libraries.

> detach tol, cache, and rng from the object

See above for RNG, otherwise we think it's a better API that way.

> support mixed continuous and discrete distribution

Mixed and Discrete distribution are definitely on the table and will be a follow up.

> The notebook example suggests that a distribution object may be instantiated with vector parameter. But the documentation gives no clue of this

More documentation will come. Now we are focusing on the basics.

> Normal._pdf_formula ... remove self and make the method a classmethod or staticmethod, or remove the distribution parameters and only access them via self

Thanks, we will have another look and consider it.

> sf to ccdf

We had some discussions here when we added features around survival analysis and this naming is less common for non experts than the complementary of the CDF.

> as compound Poisson Gamma (Tweedie).

A general note about distributions. We don't want to go back to the current status where we have more tham 100 distributions. Instead, we want to make an API which allows anyone to add their custom distributions.

We are actually thinking about making a dedicated library to host user defined distributions. We don't need to bloat the library itself and most importantly add to the maintenance pile more than we have to.

> I'd also want a way to select the distribution's sampling strategy. It may be useful to force inverse-transform, or ziggurat, or Box-Muller for the Normal distribution in particular.

There is a method parameter to do a basic method selection. But we are going to work on this more. The end goal will be to use methods which are Array API compatible so that sampling can be fast and library independent.

Something else we will look into is to see how we can integrate the work done in `stats.sampling.FastGeneratorInversion` with UNURAN. We just need to see how, here as well, it will fit with Array API support.

> cumulant support,

We can consider that in a follow up.

> One thing I haven't noticed mentioned (possible I missed it though) is which methods are required and which aren't? (when implemented a new distribution)

We have a diagram for all the logic but we need to do some small polishing. But yes, it will be in the doc.

**Comment by josef-pkt** [comment]
I think dropping the `fit` method is not a good idea:

It's the main usecase where we need to call methods with different parameters, and you need it for the design decisions.

Current distributions have many distribution specific fit methods, that are useful and took a long time to add and improve.
It's very difficult to fit some of those distributions that are not well behaved at boundaries or because of changing support.

**Comment by tupui** [comment]
Moving the `fit` method out for now does not mean that (i) it will not be added back, (ii) we cannot use any specialisation. We are considering all of that, don't worry.

**Comment by josef-pkt** [comment]
I gave up worrying and just hope for a loooong deprecation period

**Comment by WarrenWeckesser** [comment]
+1 for immutable.
+1 for implementing `fit` as a class method, where the method generates a new instance of the class.

**Comment by josef-pkt** [comment]
immutable without args in the methods kills it for any fitting applications, or we have to create a new instance each time during fitting.

(Models in statsmodels are supposed to be immutable (by user), but not having some extra args in the methods makes that very annoying in some use cases. We have some backdoor mutating in internal applications, but it looks dirty.
I just created a set of classes that have `set_params` method, but I also don't feel comfortable with it. It's too easy to mutate the wrong instance.
)

**Comment by tupui** [comment]
Yes immutability is off the table. As I said, we want to support changing parameters for things like fitting.

**Comment by tupui** [comment]
Before doing more meta-discussions around that. Please also have a look at the timings Matt put above as well. You really need to look at this with fresh eyes.

**Comment by josef-pkt** [comment]
> Would it be possible to clarify why fitting needs mutable distributions?

The main reason is overhead of instance creation, and convenience of calling a method if it has a parameter argument.

**Comment by NeilGirdhar** [comment]
> The main reason is overhead of instance creation, 

Would you happen to have a program that illustrates why this overhead might be problematic?  The reason I ask is twofold.  First, with the broadcasting interface, if you're doing a lot of fitting in parallel, then you can fit them all at once, and only one instance is created for all of the random variables.  And second, nowadays, there are plenty of mathematical JITs that would erase all of the overhead.  So, if you could show an example, we could explore if a JIT would solve your problem.

> and convenience of calling a method if it has a parameter argument.

I'm not sure why calling a method is any nicer than instantiating and calling a method on the instance?  I find the Python principle of there being "1 way to do things" to be more motivating, personally.

> Please also have a look at the timings Matt put above as well. You really need to look at this with fresh eyes.

I read (and loved) Matt's comment when I hearted it.  I see that you're referring to the timings.  I just don't see how creating instances is problematic in practice, but maybe my applications are different than other people's.

**Comment by tupui** [comment]
> I read (and loved) Matt's comment when I hearted it. I see that you're referring to the timings. I just don't see how creating instances is problematic in practice, but maybe my applications are different than other people's.

Oh I was saying that for everyone here. Plus we had a talk with you so I know you have more context from us.

What I am saying here is that if we really want to have a meaningful discussion then everyone needs to be on the same page by pulling Matt's branch locally and playing with it. Otherwise it's easy to think that it works in certain ways though it does something else entirely. Talking about overhead and such without trying it out does not make sense to me otherwise. The new API is very different from the previous one.

**Comment by fancidev** [comment]
For brainstorming purpose, Java and C# address the mutable/immutable dilemma (for String) by keeping String immutable but having a separate, mutable StringBuilder class. Do whatever you like on StringBuilder, and once you’re happy with the result, create a String from it for later use.

I feel this model may be an overkill for Python, so I don’t intend to propose it as a solution. Just for your info.

**Comment by josef-pkt** [comment]
Actually, it's difficult to find an example in statsmodels that uses scipy distribution in fit with variable parameters.

For most distributions we have explicit formulas for the loglike and its derivatives.
For loc-scale distributions without shape parameter we can just use the standard distribution on the transformed data.

The only example I currently find:
For the t distribution model with estimated df, the loglikelihood uses
`llike  = - stats.t._logpdf(errorsest/scale, df) + np_log(scale)`
which is then used with scipy.optimize.

(Models with other distributions that have shape parameters, e.g. for failure time and extreme values, are still missing in statsmodels, Numfocus thought they are not important)

So, maybe I don't need to care.

The only other time consuming loops that I remember right now are rootfinding for power and sample size computation which in many cases have to solve a cdf constraint for a shape parameter.

Most other application of scipy distribution in statsmodels are one time calls where overhead would be unimportant most of the time.
(I have a recent case with distribution.expect numerical integration inside rootfinding. But that is so slow that it's not usable for users anyway. :)

You can also time the `fit` method in scipy to see whether creating instances has a non-negligible overhead.

**Comment by josef-pkt** [comment]
@NeilGirdhar 

Switching backend or supporting backends other than numpy/scipy/pandas is currently not planned for statsmodels.
The priority is still on adding statistics and econometrics that is currently not available in Python, with reasonable performance given current tools and not on finding the fastest implementation. 
(There are several other Python packages that have parts of statsmodels' functionality but oriented on speed and performance.)

**Comment by rkern** [comment]
> We don't want to go back to the current status where we have more tham 100 distributions.

Is there a record of these policy decisions somewhere? I don't recall this being a part of the roadmap when the work got started.

**Comment by NeilGirdhar** [comment]
@josef-pkt I understand.  The reason I wanted to illustrate the Jax solution is to support my argument that I don't think that speed should be a significant consideration when making design decisions.  Whether you need to construct objects or not is totally irrelevant if you care about speed because if you care about speed, you should just use a JIT.

And just so that we're clear: I'm not sure if you meant "array library" when you said "backend"?  Because the Jax solution is a different _array library_, but it's the same _backend_ (CPU).

In short, my personal preference is for design decisions to be focused on ergonomics over efficiency since JITs provided by Array API libraries should solve the efficiency problem.

**Comment by tupui** [comment]
> > We don't want to go back to the current status where we have more tham 100 distributions.
> 
> Is there a record of these policy decisions somewhere? I don't recall this being a part of the roadmap when the work got started.

We did not make a public official announcement specific to that yet no.

While we iterated on this RFC with the other SciPy stats maintainers, it became clear for us that we did not want to leave the door open anymore to anything. The maintenance cost is just too high and none of us want to review nor maintain more than what we must.

We feel that a community effort around that would be a more sustainable solution. The new API will make it simpler to add any distributions and this does not need to live in SciPy. Hence we will push for another repository proposing a collection of distribution. I plan to quick start this repository later this year, but in the end, I and other SciPy stats maintainers will not provide any guarantee on that repo.


> And just so that we're clear: I'm not sure if you meant "array library" when you said "backend"? Because the Jax solution is a different array library, but it's the same backend (CPU).
> 
> In short, my personal preference is for design decisions to be focused on ergonomics over efficiency since JITs provided by Array API libraries should solve the efficiency problem.

Support for Array API will be added right after. The team has been quite busy adding support to stats and preparing the ground to enable a full support with the new distribution infrastructure.

**Comment by rkern** [comment]
> While we iterated on this RFC with the other SciPy stats maintainers, it became clear for us that we did not want to leave the door open anymore to anything. The maintenance cost is just too high and none of us want to review nor maintain more than what we must.

Where did this discussion take place? Not on the mailing list or the Discourse.

**Comment by rgommers** [comment]
> > Is there a record of these policy decisions somewhere? I don't recall this being a part of the roadmap when the work got started.
> 
> We did not make a public official announcement specific to that yet no.
> 
> While we iterated on this RFC with the other SciPy stats maintainers, it became clear for us that we did not want to leave the door open anymore to anything. The maintenance cost is just too high and none of us want to review nor maintain more than what we must. [...]

That's kinda not how this works. Decisions get made in public, with clear rationales. I suggest to focus on getting the new infrastructure merged without making such definitive statements on what can and cannot be contributed in the future. Deprecations/removals of existing API surface is the same. It can be mixed in one huge discussion, but I'd advise keeping it separate. It's not at all clear to me what the best way forward is for the old (very widely used infra).

The new framework seems to be really good. Let's get that completed, used, and then worry about the rest.

**Comment by tupui** [comment]
> That's kinda not how this works. Decisions get made in public, with clear rationales. I suggest to focus on getting the new infrastructure merged without making such definitive statements on what can and cannot be contributed in the future. 

To be clear, this is not a done deal in the sense that this RFC is not about that. This is just me saying that some of the current stats folks (@mdhaber, @steppi, @tirthasheshpatel, @ev-br, @chrisb83, @dschmitz89 and myself) had 6 meetings together and we felt that we did not want to go back to this state and spend our time on this in the future.

**Comment by fancidev** [comment]
> We feel that a community effort around that would be a more sustainable solution. The new API will make it simpler to add any distributions and this does not need to live in SciPy. Hence we will push for another repository proposing a collection of distribution. I plan to quick start this repository later this year, but in the end, I and other SciPy stats maintainers will not provide any guarantee on that repo.

I like this idea. I’m impressed by the large, comprehensive set of distributions in SciPy. There’s some room left for improvement as some distributions are less polished than others. A community effort could help in this regard.

**Comment by lucascolley** [comment]
For those who are following this issue but may have missed [the discourse post](https://discuss.scientific-python.org/t/enh-new-statistical-distribution-infrastructure-and-spinoff-features/1261), see gh-21050 for a first PR for the new infra.

**Comment by lorentzenchr** [comment]
>> Mixed continuous and discrete distributions are rare but do exist https://en.wikipedia.org/wiki/Tweedie_distribution . It'd make more sense to me for the distribution infrastructure to be as support-independent as possible, even down to being generic over it.
>
> IIUC @mdhaber has noted this is planned, at least I'm assuming that's what is meant by mixture distributions (and not variance scale-mixture families which is a wholely different class of distributions).

Could this be confirmed? It is important for me to know if the new infrastructure could, in the future, support such distributions with point masses such that gh-11291 could be revisited.

**Comment by mdhaber** [comment]
Yes. For example, a zero-inflated distribution could be represented as a mixture between a discrete and continuous distribution. It would be a separate class, but with a similar interface, some shared input validation code, and it would take advantage of the `ContinuousDistribution` methods for the continuous components of the mixture. You may know that methods like those for CDF, moments, and random sampling can be expressed pretty simply in terms of the CDF, moments, and random sampling methods of the underlying components (e.g. see gh-19413). That part is roughly independent of how the continuous and discrete components are implemented - although because of how `ContinuousDistribution` *is* implemented, a mixture of these and upcoming `DiscreteDistributuon`s can be more efficient than if it were composed from instances of `rv_continuous`/`rv_discrete`. I can prototype a mixture of `ContinuousDistribution`s now if it would help.

**Comment by lorentzenchr** [comment]
@mdhaber THX. Sometimes it helps to prototype a bit to have a better view if things work out as planned. But I can’t get involved too much right now.

**Comment by thobitz** [comment]
Hi,

I am working a lot with Scipy and just spent the weekend understanding why it is so difficult to change the parametrization convention of existing distributions. There are even quite well-adopted packages like paramnormal which try doing that, but they do not seem to integrate well with the scipy.stats infrastructure. The approach described below would require only implementing two functions (_parse_args and _unparse_args) in a class derived from the standard Scipy version of the distribution to translate to and from the standard SciPy convention. I thought it might be good to align first before investing further work. My Scipy version is 2.1.2. 

### Problem description
The terms “location” and “scale” do not really appear to be standard terms in statistical literature – terms like “mu” and “sigma” are equivalent for the normal distribution, but have completely different meaning in other distributions (starting even with log normal). For more complex distributions, it is unlikely to find one agreed way of parametrization (Wikipedia often lists two or three, often parameters in one convention the negative or inverse of the other). Converting conventions in application code is error-prone and hard to maintain, hence a more built-in solution may be easier to maintain.

### Solution Context
All member functions of rv_distribution apparently can be classified into 
•	Functions which take distribution parameters plus function parameters and produce some numbers like the CDF, PDF or a set of random numbers
•	Fit functions which take data, a starting point and the values of fixed parameters and return parameters of a distribution.

In principle, changing parameters should be rather trivial. In fact, there is a set of functions (_parse_args, parse_args_rvs, _parse_args_stats. _unpack_loc_scale) which read the parameter set and return the unpacked parameters. In fact, there seems to be a need for only one of these functions, as the distribution parameters are always the same.

However, it is impossible to simply specify a distribution “adopted_lognormal()” with parameters mu and sigma by overloading these _parse_args functions, as they are generated at runtime from a “parse_arg_template” and “dynamically attached” to the class instance. It is unclear which result this achieves which could not be achieved with appropriate use of *args and **kwargs; however, the disadvantage is obvious, as it blocks the standard OO mechanism of function overloading. 

On top of this, the parse_args_rvs() template even calls self._argcheck_rvs(), which is a naming completely inconsistent with the other _argcheck function, as it does not check parameters, but unpacks them into the format required by the following steps. In fact, the rvs() function subsequently calls _arg_check() to actually change the distribution parameters. 

### Solution approach
A better approach (without major changes to the infrastructure) might be
•	Have one function which parses the distribution parameters and transforms them into the parametrization expected by the scipy.stats function
•	Have one function to translate this parametrization back into the “external” format.
The flow in all functions would be
•	Take parameters, pass them to _parse_arg() to get the standard parametrization
•	Apply _argcheck() to this parameter set
•	Rescale, call normalized function (like _pdf)
•	Rescale and return result
Fit would require additional functionality to map back, and to translate fixed parameters.

### Code example 
(restore_class_params just undoes the dynamic binding; instead of three parse functions only one would be required with minor changes in the code base.)

```
import scipy.stats as st
import types

def restore_class_params(obj, params):
    if isinstance(params, str):
        params = params.split(',')
    for p in params:
        setattr(obj, p, types.MethodType(getattr(obj.__class__, p), obj))

class norm_mu_sigma_gen(st._continuous_distns.norm_gen):

    def __init__(self, name=None, *args, **kwargs):
        if name is None:
            name = 'norm_mu_sigma'
        shapes = 'mu sigma'
        super().__init__(name=name, shapes=shapes, *args, **kwargs)
        restore_class_params(self, '_parse_args,_parse_args_stats,_parse_args_rvs')

    def _parse_args(self, mu, sigma, *args, loc=0, scale=1):
        return ((), loc+mu*scale, sigma*scale)

    def _parse_args_stats(self, mu, sigma, loc=0, scale=1, moments='mv'):
        return ((), loc+mu*scale, sigma*scale, moments)

    def _parse_args_rvs(self, mu, sigma, loc=0, scale=1, size=None):
        return self._argcheck_rvs(loc+mu*scale, sigma*scale, size=size)

if __name__ == '__main__':
    norm_mu_sigma = norm_mu_sigma_gen()
    print(norm_mu_sigma.stats(mu=1, sigma=2, moments='mv'))
    print(norm_mu_sigma.pdf(0, mu=1, sigma=2))
    print(norm_mu_sigma.cdf(0, mu=1, sigma=2))
    print(norm_mu_sigma.sf(0, mu=1, sigma=2))
    print(norm_mu_sigma.ppf(0.5, mu=1, sigma=2))
    print(norm_mu_sigma.isf(0.5, mu=1, sigma=2))
    print(norm_mu_sigma.moment(2, mu=1, sigma=2))
    data = norm_mu_sigma.rvs(mu=1, sigma=2, size=1000)
    print(norm_mu_sigma.fit(data))





```

**Comment by JacobHass8** [comment]
Would it be possible to retain the frozen/unfrozen behavior by defining a static and private method where the private method calls the static method with the object's attributes. Upon an object being instantiated, the static method is then overwritten. Here's a quick example for a Gaussian distribution

```python
import numpy as np

class Gaussian():
    def __init__(self, mu, sigma):
        self.mu = mu
        self.sigma = sigma
        self.pdf = self._pdf

    @staticmethod
    def pdf(x, mu, sigma):
        return 1 / np.exp(2 * np.pi * sigma**2) * np.exp(-(x-mu)**2 / 2 / sigma**2)
    
    def _pdf(self, x):
        return Gaussian.pdf(x, self.mu, self.sigma)
```
The function `pdf` could then be calculated using a frozen or unfrozen distribution as follows
```python
# Get the pdf at locations `x`
x = np.asarray([1,0,1])

# Using a frozen distribution
norm = Gaussian(0, 1)
norm.pdf(x)

# Using an unfrozen distribution
Gaussian.pdf(x, 0, 1)
```
This would at least get rid of having to define a frozen and unfrozen object but would still require defining two functions that do the same thing.

**Comment by NeilGirdhar** [comment]
What would the benefit of that be over just writing: `Gaussian(0, 1).pdf(x)`?

Seems like having one way to do something is superior to having two ways of doing it.
